### PR TITLE
codeintel/lockfiles: Introduce yarn.lock parser

### PR DESCRIPTION
**Pull request by tsenart**
## Background

This commit introduces a yarn.lock parser and refactors the existing package-lock.json parser to reuse github.com/aquasecurity/go-dep-parser. As part of this PR, we also introduce support for sub repo permissions checking by using `git.LsFiles`.

## Test plan

Unit and integration tests.

### CODE REVIEW

**Review by efritz — APPROVED** [review]
Neat!

File: internal/codeintel/lockfiles/service.go [hunk 1]
```diff
@@ -36,15 +47,18 @@ func (s *Service) StreamDependencies(ctx context.Context, repo api.RepoName, rev
...
 		Treeish: rev,
 		Format:  "zip",
-		Paths: []string{
-			"*" + NPMFilename,
-		},
+		Paths:   paths,
 	}
 
-	rc, err := s.archiveStreamer.StreamArchive(ctx, repo, opts)
+	rc, err := s.archive(ctx, repo, opts)
```
Comment (efritz) [inline internal/codeintel/lockfiles/service.go:61]: Why do you keep undoing all the things I'd like to mock?

**Review by efritz — APPROVED** [review]

File: internal/codeintel/lockfiles/service.go (hunk 1 above)
Comment (tsenart) [reply internal/codeintel/lockfiles/service.go:61]: git.LsFiles is a package level function, and it felt a bit weird to pass in a func AND an interface when they are both related to gitserver client.

**Comment by efritz** [comment]
I'm optimistically merging to unblock other refactor efforts around the same code.

**Pull request merged**